cmd/root.go

```diff
@@ -0,0 +1,71 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"os/user"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"github.com/tlkamp/plugin-cli-example/cmd/plugins"
+	"github.com/tlkamp/plugin-cli-example/cmd/version"
+	"github.com/tlkamp/plugin-cli-example/internal/executor"
+)
+
+type PluginExecutor interface {
+	// Lookup returns true if the plugin is found by the plugin executor
+	Lookup(name string) bool
+
+	// Execute executes the plugin by the given name
+	Execute(name string, args []string, env []string) error
+}
+
+func Execute() error {
+	var (
+		pluginDIR    string
+		pluginPrefix string
+	)
+
+	rootCmd := &cobra.Command{
+		Use:   "cli-with-plugins",
+		Short: "A demo of how to create a CLI that supports arbitrary plugins",
+	}
+
+	u, err := user.Current()
+	if err != nil {
+		return fmt.Errorf("error getting current user: %w", err)
+	}
+
+	dd := strings.Join([]string{u.HomeDir, ".plugin-cli-example/plugins"}, string(os.PathSeparator))
+
+	rootCmd.PersistentFlags().StringVar(&pluginDIR, "plugin-dir", dd, "the plugin directory to use")
+	rootCmd.PersistentFlags().StringVar(&pluginPrefix, "plugin-prefix", "example", "the plugin prefix to use")
+
+	rootCmd.AddCommand(version.NewCommand())
+	rootCmd.AddCommand(plugins.NewCommand())
+
+	pieces := os.Args[1:]
+
+	if _, _, err := rootCmd.Find(pieces); err == nil {
+		// Cobra command found - execute it
+		return rootCmd.Execute()
+	}
+
+	// The command passed isn't built in, so search our plugins
+	for _, e := range []PluginExecutor{executor.New(pluginDIR, pluginPrefix)} {
+		if ok := e.Lookup(pieces[0]); !ok {
+			continue
+		}
+
+		// Pugin found - execute it
+		if err := e.Execute(pieces[0], pieces[1:], os.Environ()); err != nil {
+			fmt.Fprintf(os.Stderr, "error executing plugin: %v", err)
+			os.Exit(1)
+		}
+
+		os.Exit(0)
```

Comment:
discussion: the way I'm reading this is that it will exit 0 if the user types a command that is neither a built-in nor a plugin. If my read is accurate, I don't think the CLI should exit 0.